kong/tools/uri.lua

```diff
@@ -12,52 +9,108 @@ local table_concat = table.concat
 local ngx_re_find = ngx.re.find
 local ngx_re_gsub = ngx.re.gsub
 
+local table_new = table.new
+
+
+-- Charset:
+--   reserved = "!" / "*" / "'" / "(" / ")" / ";" / ":" / "@" / "&" / "=" / "+" / "$" / "," / "/" / "?" / "%" / "#" / "[" / "]"
+--   unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+--   other: * (meaning any char that is not mentioned above)
+
+-- Reserved characters have special meaning in URI. Encoding/decoding it affects the semantics of the URI;
+-- Unreserved characters are safe to use as part of HTTP message without encoding;
+-- Other characters has not special meaning but may be not safe to use as part of HTTP message without encoding;
+
+-- We should not unescape or escape reserved characters;
+-- We should unescape but not escape unreserved characters;
+-- We choose to unescape when processing and escape when forwarding for other characters
+
+local RESERVED_CHARS = "!*'();:@&=+$,/?%#[]"
+
+local HYPHEN_BYTE = string_byte('-')
+local DOT_BYTE = string_byte('.')
+local UNDERSCORE_BYTE = string_byte('_')
+local TILDE_BYTE = string_byte('~')
+local CAP_A_BYTE = string_byte('A')
+local CAP_Z_BYTE = string_byte('Z')
+local A_BYTE = string_byte('a')
+local Z_BYTE = string_byte('z')
+local ZERO_BYTE = string_byte('0')
+local NINE_BYTE = string_byte('9')
+
+local CHAR_RESERVED = true
+local CHAR_UNRESERVED = false
+local CHAR_OTHERS = nil -- luacheck: ignore
+
+local char_urlencode_type = table_new(256, 0) do
+  -- reserved
+  for i = 1, #RESERVED_CHARS do
+    char_urlencode_type[string_byte(RESERVED_CHARS, i)] = CHAR_RESERVED
+  end
+
+  -- unreserved
+  for num = A_BYTE, Z_BYTE do
+    char_urlencode_type[num] = CHAR_UNRESERVED
+  end
+
+  for num = CAP_A_BYTE, CAP_Z_BYTE do
+    char_urlencode_type[num] = CHAR_UNRESERVED
+  end
+
+  for num = ZERO_BYTE, NINE_BYTE do
+    char_urlencode_type[num] = CHAR_UNRESERVED
+  end
+
+  for _, num in ipairs{
+    HYPHEN_BYTE, DOT_BYTE, UNDERSCORE_BYTE, TILDE_BYTE,
+  } do
+    char_urlencode_type[num] = CHAR_UNRESERVED
+  end
```

Comment:
I actually think just writing:

```suggestion
    char_urlencode_type[HYPHEN_BYTE] = CHAR_UNRESERVED
    char_urlencode_type[DOT_BYTE] = CHAR_UNRESERVED
```

is better. It is even less lines of code.